internal/controller/apiextensions/composite/reconciler.go

```diff
@@ -423,6 +426,25 @@ func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error)
 		if rdy {
 			ready++
 		}
+
+		if err := r.composite.Render(ctx, cr, cd, comp.Spec.Resources[i]); err != nil {
+			log.Debug(errRenderCR, "error", err)
+			r.record.Event(cr, event.Warning(reasonCompose, err))
+			return reconcile.Result{RequeueAfter: shortWait}, nil
+		}
+	}
+
+	// Update composite status and spec with patches from composed resources
+	if err := r.client.Status().Update(ctx, cr); err != nil {
+		log.Debug(errUpdate, "error", err)
+		r.record.Event(cr, event.Warning(reasonCompose, err))
+		return reconcile.Result{RequeueAfter: shortWait}, nil
+	}
+
+	if err := r.client.Update(ctx, cr); err != nil {
+		log.Debug(errUpdate, "error", err)
+		r.record.Event(cr, event.Warning(reasonCompose, err))
+		return reconcile.Result{RequeueAfter: shortWait}, nil
```

Comment:
You might need to test that this is working as expected. We've often had issues in the past with updating status and then updating the rest of an object. If I recall correctly this is because when you update the status the controller-runtime client loads the updated object from the API server, thus reverting any pending changes to the object's spec or metadata. :(

This is kind of a pain, because it might mean we can't neatly render out the new composite resource and then update both its spec and status. We may instead need to render the spec, persist that, then do the status, then persist that (or vice versa).

Reply:
I'll verify again with a debugger that this is being a persisted, but I thought I recall the proper order being status update, then update. The update itself I believe reverts status.

Reply:
I'll definitely be happy if I'm wrong - I like how the code is laid out now.

CC @muvaf you usually remember the details of this quirk better than I do.

Reply:
@negz your assessment is correct. Any update operation returns the object from the api-server and unmarshalls it into the object you passed, in this case `cr`. What we usually do is to issue a spec update if there is a change in spec and only then update the status.

Reply:
Yeah, after testing it seems either update, spec or status, resets the other sub-field to values from the api-server. In this case updating status first works properly because each sub-field is being set first. So status is set then updated, then spec is set and updated.

Reply:
@mcavoyk Apologies if I'm belabouring the point here, but I'm still struggling to grok how this is working.

Reading the code I would expect:

1. We iterate over all composed resources calling `r.composite.Render` which potentially patches from the composed resource to either the status or the spec of the composite resource (i.e. spec fields that a claim - if any - will allow us to late init).
2. We update the status, which resets any pending spec changes from 1.
3. Any pending spec changes were reset in 2, so this is always a no-op.

What am I missing?

Reply:
My mistake, I thought this was reviewing the section in the claim controller which updates status and spec of the claim. Yes from looking at this, it must be only working in my tests after multiple reconciles. This should be updated to preserve the values in status, update the spec, then update status instead of the current logic.

Reply:
This should be resolved now, one of the reasons I did not notice this was spec propagated in my testing was from the base composed, and not a later initialization of the composed. Before, spec updates to composite were not applied and lost when the status was updated.